mariamihaelafainarea/assignment4/Assignment4

```diff
@@ -0,0 +1,78 @@
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package assignment4;
+
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ *
+ * @author maria
+ */
+public class Assignment4 {
+
+    int n,m;
+    boolean[][] matrix;
+    DijointSet d;
+    //false - water
+    //true - land
+    //Please return number of islands ! 
+
+    public Assignment4(int n, int m, boolean[][] matrix) {
+        this.n = n;
+        this.m = m;
+        this.matrix = matrix;
+    }
+    //First, assign for each cell a different set
+    public void putIntoDifferentSets() {
+        d = new DijointSet();
+        for (int i = 0; i < n*m; i++) {
+            d.makeSet(i);
+        }
+    }
+    //Then, the flood fill alg will get together
+    //the cells in the neighbourhood => set of islands
+    public void applyFloodFillOnMe(AtomicBoolean[][] verified) {
+        FloodFill f = new FloodFill();
+        for (int i = 0; i < n ;i++) {
+            for(int j = 0; j < m; j++) {
+                f.floodFill(d, verified, matrix, n, m,i,j);
+            }
+        }
+    
+    }
+    //to obtain the number of islands
+    //we get the parent nodes of all sets 
+    //that were processed.
+    //if the node isn't verrified there
+    //is water, not land
+    public int getNrOfDifferentSetsWithIslands(AtomicBoolean[][] verified) {
+        Set<Integer> x = new TreeSet<>();
+        for (int i = 0; i < n ; i++) {
+            for(int j = 0; j < m; j++) {
+                if(verified[i][j].get()) {
+                    //only if i have put it into a set! 
+                    //otherwise it is not a land
+                    x.add(d.findSet(i*m + j));
+                }
+            } 
+        }
+        return x.size();
+    }
+    public int findNumberOfIslandsProcedure() {
+        putIntoDifferentSets();
+        AtomicBoolean[][] verified = new AtomicBoolean[n][m];
```

Comment:
I'm pretty sure this time you can get this done without using AtomicBooleans :)

Reply:
Yes. it works apparently. But i do not understand. Why? I can see a difference between this and Assignment 2 : there were 2 parameters set into the function and here is just a matrix. I can only think about that in Assignment 2 if there were 2 threads maybe somehow the set procedure would be broken in 2 parts. Do you have any idea?